control/grpc.py

```diff
@@ -416,112 +438,781 @@ def delete_subsystem_safe(self, request, context=None):
-                    if not request.nsid:
-                        request.nsid = nsid
+                    namespace_add_req = pb2.namespace_add_req()
+                    try:
+                        namespace_add_req.rbd_pool_name=ns_entry["rbd_pool_name"]
+                    except KeyError:
+                        pass
+                    try:
+                        namespace_add_req.rbd_image_name=ns_entry["rbd_image_name"]
+                    except KeyError:
+                        pass
+                    try:
+                        namespace_add_req.subsystem_nqn=ns_entry["subsystem_nqn"]
+                    except KeyError:
+                        pass
+                    try:
+                        namespace_add_req.nsid=ns_entry["nsid"]
+                    except KeyError:
+                        pass
+                    try:
+                        namespace_add_req.block_size=ns_entry["block_size"]
+                    except KeyError:
+                        pass
+                    try:
+                        namespace_add_req.uuid=ns_entry["uuid"]
+                    except KeyError:
+                        pass
+                    namespace_add_req.anagrpid=request.anagrpid
                     json_req = json_format.MessageToJson(
-                        request, preserving_proto_field_name=True, including_default_value_fields=True)
-                    self.gateway_state.add_namespace(request.subsystem_nqn,
-                                                     str(nsid), json_req)
+                            namespace_add_req, preserving_proto_field_name=True, including_default_value_fields=True)
+                    self.gateway_state.add_namespace(request.subsystem_nqn, nsid, json_req)
                 except Exception as ex:
-                    self.logger.error(
-                        f"Error persisting add_namespace {nsid}: {ex}")
-                    raise
+                    errmsg = f"Error persisting change load balancing group for namespace {nsid_msg}in {request.subsystem_nqn}:\n{ex}"
+                    self.logger.error(errmsg)
+                    return pb2.req_status(status=errno.EINVAL, error_message=errmsg)
 
-        return pb2.nsid_status(nsid=nsid, status=True)
+        return pb2.req_status(status=0, error_message=os.strerror(0))
 
-    def add_namespace(self, request, context=None):
-        return self.execute_grpc_function(self.add_namespace_safe, request, context)
+    def namespace_change_load_balancing_group(self, request, context=None):
+        """Changes a namespace load balancing group."""
+        return self.execute_grpc_function(self.namespace_change_load_balancing_group_safe, request, context)
 
-    def remove_namespace_safe(self, request, context=None):
+    def remove_namespace_from_state(self, nqn, nsid, context):
+        if not context:
+            return pb2.req_status(status=0, error_message=os.strerror(0))
+
+        if context:
```

Comment:
nit: if context: is not required, due to condition in line 752

Reply:
Oops... I copy and pasted this in several places and didn't notice that I have the condition right above it. Will fix that.